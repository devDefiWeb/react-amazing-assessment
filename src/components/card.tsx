import styled from "styled-components";

const Card = styled.div`
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
    transition: 0.3s;
    border-radius: 5px;
    width: 35%;
    margin: auto;
    background-color: #F1F5F8;
    margin-top: 100px;
    margin-bototm: 100px;
    :hover {
        box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
    }
`

export default Card;

export const CardBody = styled.div`
    padding: 30px;
`

export const CardHeader = styled.div`
    border-bottom: solid 1px #cfcfcf;
    padding: 20px;
`
