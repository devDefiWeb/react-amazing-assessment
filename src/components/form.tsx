import styled from "styled-components";

export const Title = styled.div`
    font-size: 40px;
    text-align: center;
    font-weight: center;
`

export const Row = styled.div`
    display: flex;
    align-items: stretch;
    width: 100%;

    @media (max-width: 800px) {
        flex-direction: column;
    }
`;

export const Item = styled.div`
    margin: 5px;
    text-align: center;
    line-height: 40px;
    font-size: 30px;
    width: 100%;
`

export const StyledInput = styled.input`
    width: 100%;
    height: 40px;
    padding: 10px;
    box-sizing: border-box;
`

export const StyledLabel = styled.label`
    float: left;
    font-size: 18px;
    margin-bottom: -10px;
    font-weight: 500;
`

export const StyledSelect = styled.select`
    padding: 10px;
    margin: auto;
    height: 40px;
    width: 100%;
    margin-top: 10px;
`

export const StyledTextarea = styled.textarea`
    margin-top: 10px;
    width: 100%;
    height: 100px;
    resize: vertical;
    box-sizing: border-box;
    padding: 10px;
`

export const SubmitButton = styled.button`
    width: 150px;
    height: 40px;
    background-color: #006BA8;
    color: white;
    border-width: 0px;
    float: right;
    margin-top: 10px;

    :hover {
        background-color: #065b8b;
        cursor: pointer;
    }
`

export const StyledSpan = styled.span`
    color: #ed8888;
`
