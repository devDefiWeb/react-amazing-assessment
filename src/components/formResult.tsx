import React from "react";
import { useDispatch, useSelector } from "react-redux";
import BlockUi from 'react-block-ui';
import { useNavigate } from 'react-router-dom'
import { clearFieldSetData } from "../redux/actions/form.action";
import { IAppState } from "../redux/reducers";
import { Field } from "../model";
import Card, { CardBody, CardHeader } from "./card";
import { Item, StyledLabel, StyledInput, Row, Title, SubmitButton } from "./form";

const FormResult: React.FC = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();

    const { isLoading, data: fieldSetData } = useSelector((store: IAppState) => store.form);

    const handleClickReturnButton = () => {
        dispatch(clearFieldSetData());
        navigate('/submit');
    }

    const getComponentFromFieldData = (fieldData: Field) => {
        switch (fieldData.type) {
            case "text":
                return (
                    <Item key={fieldData.id}>
                        <StyledLabel>
                            {fieldData.id} :
                        </StyledLabel>
                        <StyledInput
                            value={fieldData.value}
                            readOnly
                        />
                    </Item>
                );

            case "select":
                return (
                    <Item key={fieldData.id}>
                        <StyledLabel>
                            {fieldData.id} :
                        </StyledLabel>
                        <StyledInput
                            value={fieldData.value}
                            readOnly
                        />
                    </Item>
                );

            case "textarea":
                return (
                    <Item key={fieldData.id}>
                        <StyledLabel>
                            {fieldData.id} :
                        </StyledLabel>
                        <StyledInput
                            value={fieldData.value}
                            readOnly
                        />
                    </Item>
                )
        }
    }
    const fieldList = fieldSetData.map(item => {
        if (Array.isArray(item)) {
            return (
                <Row key={item[0].id}>
                    {item.map(dataField => getComponentFromFieldData(dataField))}
                </Row>
            );
        } else {
            return (
                <Row key={item.id}>
                    {getComponentFromFieldData(item)}
                </Row>
            );
        }
    });

    return (
        <BlockUi blocking={isLoading}>
            <Card>
                <CardHeader>
                    <Title>
                        Thank you!
                    </Title>
                </CardHeader>
                <CardBody>
                    {fieldList}
                    <Row>
                        <Item>
                            <SubmitButton onClick={() => handleClickReturnButton()}>
                                Return
                            </SubmitButton>
                        </Item>
                    </Row>
                </CardBody>
            </Card>
        </BlockUi>
    );
}

export default FormResult;
