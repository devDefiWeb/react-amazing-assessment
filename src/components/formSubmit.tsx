import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import useDynamicRefs from 'use-dynamic-refs';
import BlockUi from 'react-block-ui';
import { useNavigate } from 'react-router-dom'
import { getFieldSetData as getFieldSetData, updateFieldSetData } from "../redux/actions/form.action";
import { IAppState } from "../redux/reducers";
import { Field, FieldSetData } from "../model";
import Card, { CardBody, CardHeader } from "./card";
import { Item, StyledLabel, StyledInput, StyledSelect, StyledTextarea, Row, Title, SubmitButton, StyledSpan } from "./form";
import { isValidEmailString, isValidPhoneString } from "../util";

const FormSubmit: React.FC = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();
    const [getRef, setRef] = useDynamicRefs();

    // state from redux store
    const { isLoading, data } = useSelector((store: IAppState) => store.form);

    // component state
    const [fieldSetData, setFieldSetData] = useState<FieldSetData>([]);

    // fetch fieldSetData
    useEffect(() => {
        dispatch(getFieldSetData());
    }, []);

    // store fieldSetData to component state 
    useEffect(() => {
        setFieldSetData(data);
    }, [data]);

    const updateData = (key: string, value: string) => {
        let _fieldSetData = fieldSetData.slice();

        _fieldSetData.map(item => {
            if (Array.isArray(item)) {
                item.map(field => {
                    if (field.id === key) {
                        field.value = value;
                    }
                });
            } else {
                if (item.id === key) {
                    item.value = value;
                }
            }
        });

        setFieldSetData(_fieldSetData);
    }

    const focusField = (fieldId: string) => {
        let ref = getRef(fieldId);
        if (ref !== undefined) {
            (ref.current as any).focus();
        }
    }

    const isValidateForm = (): boolean => {
        let fieldSetList: Field[] = fieldSetData.reduce((res: Field[], item: Field | Field[]) => (
            Array.isArray(item) ? [...res, ...item] : [...res, item]
        ), []);

        for (let i = 0; i < fieldSetList.length; i++) {
            let fieldData = fieldSetList[i];
            const fieldId = fieldData.id.toLowerCase();
            if (fieldId.indexOf("email") >= 0) {
                if (fieldData.value === "" && fieldData.required) {
                    alert(`Field ${fieldData.id} is required!`);
                    focusField(fieldData.id);
                    return false;
                }
                else if (fieldData.value !== undefined && fieldData.value !== "" && !isValidEmailString(fieldData.value)) {
                    alert(`Invalid Field ${fieldData.id}!`);
                    focusField(fieldData.id);
                    return false;
                }
            }
            if (fieldId.indexOf("phone") >= 0) {
                if (fieldData.value === "" && fieldData.required) {
                    alert(`Field ${fieldData.id} is required!`);
                    focusField(fieldData.id);
                    return false;
                }
                else if (fieldData.value !== undefined && fieldData.value !== "" && !isValidPhoneString(fieldData.value)) {
                    alert(`Invalid Field ${fieldData.id}!`);
                    focusField(fieldData.id);
                    return false;
                }
            }
            if (fieldData.required && (fieldData.value === "" || fieldData.value === undefined)) {
                alert(`Field ${fieldData.id} is required!`);
                focusField(fieldData.id);
                return false;
            }
        }

        return true;
    }

    const handleClickSubmitButton = () => {
        if (isValidateForm()) {
            dispatch(updateFieldSetData(fieldSetData));
            navigate('/result');
        }
    }

    const getComponentFromFieldData = (fieldData: Field) => {
        switch (fieldData.type) {
            case "text":
                return (
                    <Item key={fieldData.id}>
                        <StyledLabel>
                            {(fieldData.required) ? (
                                <StyledSpan>
                                    *&nbsp;
                                </StyledSpan>
                            ) : null}
                            {fieldData.id} :
                        </StyledLabel>
                        <StyledInput
                            required={fieldData.required}
                            name={fieldData.id}
                            placeholder={fieldData.placeholder}
                            onChange={(e: React.ChangeEvent<HTMLInputElement>) => updateData(fieldData.id, e.target.value)}
                            ref={setRef(fieldData.id) as React.RefObject<HTMLInputElement>}
                        />
                    </Item>
                );

            case "select":
                return (
                    <Item key={fieldData.id}>
                        <StyledLabel>
                            {fieldData.id} :
                        </StyledLabel>
                        <StyledSelect
                            defaultValue=""
                            onChange={(e: React.ChangeEvent<HTMLSelectElement>) => updateData(fieldData.id, e.target.value)}
                            ref={setRef(fieldData.id) as React.RefObject<HTMLSelectElement>}
                        >
                            <option value=""></option>
                            {fieldData.options?.map((optionName, index) =>
                                <option value={optionName} key={index}>{optionName}</option>
                            )}
                        </StyledSelect>
                    </Item>
                );

            case "textarea":
                return (
                    <Item key={fieldData.id}>
                        <StyledLabel>
                            {fieldData.id} :
                        </StyledLabel>
                        <StyledTextarea
                            placeholder={fieldData.placeholder}
                            onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => updateData(fieldData.id, e.target.value)}
                            ref={setRef(fieldData.id) as React.RefObject<HTMLTextAreaElement>}
                        />
                    </Item>
                )
        }
    }

    const fieldList = fieldSetData.map(item => {
        if (Array.isArray(item)) {
            return (
                <Row key={item[0].id}>
                    {item.map(dataField => getComponentFromFieldData(dataField))}
                </Row>
            );
        } else {
            return (
                <Row key={item.id}>
                    {getComponentFromFieldData(item)}
                </Row>
            );
        }
    });

    return (
        <BlockUi blocking={isLoading}>
            <Card>
                <CardHeader>
                    <Title>
                        Application Form
                    </Title>
                </CardHeader>
                <CardBody>
                    {fieldList}
                    <Row>
                        <Item>
                            <SubmitButton onClick={() => handleClickSubmitButton()}>
                                Submit
                            </SubmitButton>
                        </Item>
                    </Row>
                </CardBody>
            </Card>
        </BlockUi>
    );
}

export default FormSubmit;
