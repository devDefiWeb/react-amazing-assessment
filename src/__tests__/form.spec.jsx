import renderer from 'react-test-renderer'
import React from 'react'
import { shallow } from 'enzyme'
import { Title, Row, Item, StyledInput, StyledLabel, StyledSelect, StyledTextarea, SubmitButton, StyledSpan } from '../components/form'

describe('<Title />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<Title />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<Row />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<Row />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<Item />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<Item />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<StyledInput />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<StyledInput />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<StyledLabel />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<StyledLabel />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<StyledSelect />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<StyledSelect />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<StyledTextarea />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<StyledTextarea />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<SubmitButton />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<SubmitButton />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('<StyledSpan />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(<StyledSpan />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
