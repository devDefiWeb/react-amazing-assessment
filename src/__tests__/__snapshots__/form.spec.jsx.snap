// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> Should render correctly 1`] = `
<div
  className="sc-dkrFOg fxVuLP"
/>
`;

exports[`<Row /> Should render correctly 1`] = `
<div
  className="sc-gswNZR fRojqK"
/>
`;

exports[`<StyledInput /> Should render correctly 1`] = `
<input
  className="sc-hLBbgP icvjiK"
/>
`;

exports[`<StyledLabel /> Should render correctly 1`] = `
<label
  className="sc-eDvSVe TLawV"
/>
`;

exports[`<StyledSelect /> Should render correctly 1`] = `
<select
  className="sc-jSUZER epeykz"
/>
`;

exports[`<StyledSpan /> Should render correctly 1`] = `
<span
  className="sc-ftTHYK facPMY"
/>
`;

exports[`<StyledTextarea /> Should render correctly 1`] = `
<textarea
  className="sc-gKPRtg kqGUpY"
/>
`;

exports[`<SubmitButton /> Should render correctly 1`] = `
<button
  className="sc-iBYQkv ihheyn"
/>
`;

exports[`<TextInput /> Should render correctly 1`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<TextInput /> Should render correctly 2`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<TextInput /> Should render correctly 3`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<TextInput /> Should render correctly 4`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<TextInput /> Should render correctly 5`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<TextInput /> Should render correctly 6`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<TextInput /> Should render correctly 7`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;

exports[`<Title /> Should render correctly 1`] = `
<div
  className="sc-bcXHqe phYlm"
/>
`;
